File: python/day-01.py
#!/usr/bin/env/python3
import numpy as np

from common import get_input_data, pretty_duration_ns, time_function, check_result

DAY = "01"
SOLUTION_PART_1 = "2344935"
SOLUTION_PART_2 = "27647262"


# region Part 1
def solve_part_1(input_data: str) -> str:
    data = np.array([list(map(int, line.split())) for line in input_data.splitlines()])

    left = np.sort(data[:, 0])
    right = np.sort(data[:, 1])

    result = np.sum(np.abs(left - right))
    return f"{result}"


# enregion


# region Part 2
def solve_part_2(input_data: str) -> str:
    data = np.array([list(map(int, line.split())) for line in input_data.splitlines()])

    left = np.sort(data[:, 0])
    right = np.sort(data[:, 1])

    unique, counts = np.unique(right, return_counts=True)
    occurences = dict(zip(unique.tolist(), counts.tolist()))

    result = sum(num * occurences.setdefault(num, 0) for num in left)

    return f"{result}"


# endregion


# region Main
def main():
    input_data, io_duration = time_function(get_input_data, day=DAY)
    print(f"Input data loaded in {pretty_duration_ns(io_duration)}")

    answer_part_1, part1_duration = time_function(solve_part_1, input_data=input_data)
    print(f"Part 1: {answer_part_1} (solved in {pretty_duration_ns(part1_duration)})")
    check_result(answer_part_1, SOLUTION_PART_1)

    answer_part_2, part2_duration = time_function(solve_part_2, input_data=input_data)
    print(f"Part 2: {answer_part_2} (solved in {pretty_duration_ns(part2_duration)})")
    check_result(answer_part_2, SOLUTION_PART_2)


# endregion

if __name__ == "__main__":
    main()

File: rust/src/bin/day_01.rs
use anyhow::*;
use aoc2024::*;
use std::collections::HashMap;

const DAY: &str = "01";
const SOLUTION_PART_1: &str = "2344935";
const SOLUTION_PART_2: &str = "27647262";

fn input_to_vecs(input: &str) -> (Vec<u32>, Vec<u32>) {
    let mut left = Vec::<u32>::new();
    let mut right = Vec::<u32>::new();

    for line in input.lines() {
        if !line.is_empty() {
            let mut elems = line.split_whitespace();
            left.push(elems.next().unwrap().parse::<u32>().unwrap());
            right.push(elems.next().unwrap().parse::<u32>().unwrap());
        }
    }

    left.sort();
    right.sort();

    (left, right)
}

//region Part 1

fn solve_part_1(input_data: &str) -> Result<String> {
    let (left, right) = input_to_vecs(input_data);

    let mut total = 0;
    for i in 0..left.len() {
        total += left[i].abs_diff(right[i]);
    }
    Ok(format!("{}", total))
}
//endregion

//region Part 2

fn solve_part_2(input_data: &str) -> Result<String> {
    let (left, right) = input_to_vecs(input_data);

    let mut presents_in_right = HashMap::<u32, u32>::new();
    for elem in right.iter() {
        let tmp = presents_in_right.get(elem).unwrap_or(&0);
        presents_in_right.insert(*elem, *tmp + 1);
    }

    let mut result = 0;
    for elem in left {
        result += elem * presents_in_right.get(&elem).unwrap_or(&0);
    }
    Ok(format!("{}", result))
}
//endregion

fn main() -> Result<()> {
    start_day(DAY);

    let (input_data, duration) = time_function!(get_input_data(DAY)?);
    println!("Input data loaded in {}", pretty_duration(duration));

    let (answer_part_1, part1_duration) = time_function!(solve_part_1(&input_data)?);
    println!(
        "Part 1: {} (solved in {})",
        answer_part_1,
        pretty_duration(part1_duration)
    );
    check_result(&answer_part_1, SOLUTION_PART_1);

    let (answer_part_2, part2_duration) = time_function!(solve_part_2(&input_data)?);
    println!(
        "Part 2: {} (solved in {})",
        answer_part_2,
        pretty_duration(part2_duration)
    );
    check_result(&answer_part_2, SOLUTION_PART_2);

    Ok(())
}
